### ISSUE TITLE
RFC: Organize everything around the `infinity` service

### ISSUE THREAD

**Issue by hhff** [ember-infinity/ember-infinity#284]
Ember Infinity I think actually predates Ember services. Now that services are in place, and we're already using one, I think there's a compelling argument to move away from the route-centric approach and organize Ember Infinity around the `infinity` service. Some examples:

## The old way
```
Route.extend(InfinityRoute, {
  model() {
    return this.infinityModel('products', { perPage: 12 });
  }
});
```

```
{{infinity-loader model=model}}
```

This creates a coupling between the collection assembled in the model hook, and the template. The `infinity-loader` depends on a connection between things happening in the route, and things happening in the template (the `infinityLoad` action, responding to new fetches, etc).

Historically, this has caused problems with:
- Nested Infinity Routes (two scrollable collections on separate routes)...
- So we moved to closure actions / route-action helper.  However, now the route-action helper is giving us trouble in Ember 3.0+
- Caching across page reloads
- Model refreshing

## A new suggestion

```
Route.extend(InfinityRoute, {
  model() {
    return this.infinityModel('products', { perPage: 12, collectionLabel: 'masterProductsCollection' });
  }
});
```

```
{{infinity-loader collection='masterProductsCollection'}}
```

In addition, I'd suggest deprecating the `InfinityRoute` and instead suggesting this pattern:

```
Route.extend({
  infinity: service(),

  model() {
    return this.get('infinity').model('products', { perPage: 12, collectionLabel: 'masterProductsCollection' });
  }
});
```

In this format, we'll have these pros:
- `infinity-loader` injects the `infinity` service, and makes calls directly to it, rather than through route actions
- the `infinity-loader` will keep reference to collections that are already loaded, as they'll be cached on the `infinity` service under a unique label
- Nested routes with "just work"

Cons:
- using Ember's `route#reload` probably won't work here as I think we should design `infinity#model` to return a cached collection if it already has it

## Is this backwards compatible

yes, it can be!
- We can use the existing `InfinityRoute` mixin to call out to the infinity service instead of doing it's magic on the route level.
- We can allow for a model property in `infinity-loader`, and do a deprecation warning when it's not `null`.

**Comment by snewcomer** [comment]
I think one of the more difficult items to deal with here is what to do with the `store`.  In many cases, the magic of having the `store` living on the route allowed us to assume the store was available, and if the user wanted a custom store, they could pass it in the options hash to the `infinityModel` hook.  

Now, with a `service` + `component` they are a lot more isolated, so we need to pass the store, most likely to the service as a required argument to the options hash.

```
  model() {
    return this.get('infinity').model('products', { perPage: 12, collectionLabel: 'masterProductsCollection', store: this.store }); 
  }
```

We could also do some type of initialization work and find it with `getOwner` in the service, unless the store is passed into the service as an argument.  @hhff What do you think?

**Comment by davidbilling** [comment]
@snewcomer Just to inject the store in the component with "store: inject()," is no good?

**Comment by davidbilling** [comment]
@snewcomer I agree, first check for store property from infinity model, If users have custom store it should be set there then. I can't come up with any other ideas right now...
